enterprise/ha/src/main/java/org/neo4j/kernel/ha/com/master/ConversationManager.java

```diff
@@ -110,12 +127,22 @@ public void end( RequestContext context )
 
     public Set<RequestContext> getActiveContexts()
     {
-        return conversations.keys();
+        return conversations != null ? conversations.keys() : Collections.<RequestContext>emptySet() ;
     }
 
-    public void remove( RequestContext context )
+    /**
+     * Stop conversation for specified context.
+     * Conversation will still hold all already acquired locks, but will release all waiters and it will be
+     * impossible to get new locks out of it.
+     * @param context - context for which conversation should be stopped
+     */
+    public void stop( RequestContext context )
     {
-        conversations.remove( context );
+        Conversation conversation = conversations.end( context );
+        if ( conversation != null && conversation.isActive() )
+        {
+            conversation.stop();
+        }
```

Comment:
Instead, I think you should give the `TimedRepository` the ability to take a passivation function, which is eagerly called on `end` and strictly ordered before the call to the reaper. Then you can put this code in the passivation function.